addons/point_of_sale/static/src/js/Screens/OrderManagementScreen/ControlButtons/InvoiceButton.js

```diff
@@ -0,0 +1,159 @@
+    'use strict';
+
+    const { useListener } = require('web.custom_hooks');
+    const { useContext } = owl.hooks;
+    const { isRpcError } = require('point_of_sale.utils');
+    const PosComponent = require('point_of_sale.PosComponent');
+    const OrderManagementScreen = require('point_of_sale.OrderManagementScreen');
+    const OrderFetcher = require('point_of_sale.OrderFetcher');
+    const Registries = require('point_of_sale.Registries');
+    const contexts = require('point_of_sale.PosContext');
+
+    class InvoiceButton extends PosComponent {
+        constructor() {
+            super(...arguments);
+            useListener('click', this._onClick);
+            this.orderManagementContext = useContext(contexts.orderManagement);
+        }
+        get selectedOrder() {
+            return this.orderManagementContext.selectedOrder;
+        }
+        set selectedOrder(value) {
+            this.orderManagementContext.selectedOrder = value;
+        }
+        get isAlreadyInvoiced() {
+            if (!this.selectedOrder) return false;
+            return Boolean(this.selectedOrder.account_move);
+        }
+        get commandName() {
+            if (!this.selectedOrder) {
+                return 'Invoice';
+            } else {
+                if (this.selectedOrder.isFromClosedSession) return 'Cannot Invoice';
+                return this.isAlreadyInvoiced ? 'Reprint Invoice' : 'Invoice';
+            }
+        }
+        get isHighlighted() {
+            return this.selectedOrder && !this.isAlreadyInvoiced && !this.selectedOrder.isFromClosedSession;
+        }
+        async _downloadInvoice(orderId) {
+            try {
+                await this.env.pos.do_action('point_of_sale.pos_invoice_report', {
+                    additional_context: {
+                        active_ids: [orderId],
+                    },
+                });
+            } catch (error) {
+                if (error instanceof Error) {
+                    throw error;
+                } else {
+                    // NOTE: error here is most probably undefined
+                    this.showPopup('ErrorPopup', {
+                        title: this.env._t('Network Error'),
+                        body: this.env._t('Unable to download invoice.'),
+                    });
+                }
+            }
+        }
+        async _invoiceOrder() {
+            const order = this.selectedOrder;
+            if (!order) return;
+
+            const orderId = order.backendId;
+
+            // Part 0.1. If already invoiced, print the invoice.
+            if (this.isAlreadyInvoiced) {
+                await this._downloadInvoice(orderId);
+                return;
+            }
+
+            // Part 0.2. Check if order belongs to an active session.
+            // If not, do not allow invoicing.
+            const isSessionClosed = await this.rpc({
+                model: 'pos.order',
+                method: 'check_if_session_closed',
```

Comment:
So we need the "check_if_session_closed"? we have a "is_session_closed" field on the order. And if the session is closed in the meantime, the error could be returned by the call to 'write' and/or the call to 'action_pos_order_invoice'.

Reply:
This information is already in the models.Order thru isFromClosedSession. Indeed this rpc call is not needed.